src/Validation/URLValidationProvider.php

```diff
@@ -0,0 +1,216 @@
+	 *         @type int $total The total number of URLs for this type, valid or invalid.
+	 *     }
+	 * }
+	 */
+	public $validity_by_type = [];
+
+	/**
+	 * Locks validation.
+	 */
+	private function lock() {
+		set_transient( self::LOCK_TRANSIENT, 'locked', $this->get_lock_timeout() );
+	}
+
+	/**
+	 * Unlocks validation.
+	 */
+	private function unlock() {
+		delete_transient( self::LOCK_TRANSIENT );
+	}
+
+	/**
+	 * Returns whether validation is currently locked.
+	 *
+	 * @return boolean
+	 */
+	public function is_locked() {
+		return get_transient( self::LOCK_TRANSIENT );
+	}
+
+	/**
+	 * Provides the length of time, in seconds, to lock validation when this runs.
+	 *
+	 * @return int
+	 */
+	private function get_lock_timeout() {
+		/**
+		 * Filters the length of time to lock URL validation when a process starts.
+		 *
+		 * @param int $timeout Time in seconds. Default 300 seconds.
+		 */
+		return apply_filters( 'amp_validation_lock_timeout', 5 * MINUTE_IN_SECONDS );
+	}
+
+	/**
+	 * Runs a callback with a lock set for the duration of the callback.
+	 *
+	 * @param callable $callback Callback to run with the lock set.
+	 * @return mixed  WP_Error if a lock is in place. Otherwise, the result of the callback or void if it doesn't return anything.
+	 */
+	public function with_lock( $callback ) {
+		if ( $this->is_locked() ) {
+			return new WP_Error(
+				'amp_url_validation_locked',
+				__( 'URL validation cannot start right now because another process is already validating URLs. Try again in a few minutes.', 'amp' )
+			);
+		}
+
+		$this->lock();
+		$result = call_user_func( $callback );
+		$this->unlock();
+
+		return $result;
+	}
+
+	/**
+	 * Validates a URL, stores the results, and increments the counts.
+	 *
+	 * @param string $url  The URL to validate.
+	 * @param string $type The type of template, post, or taxonomy.
+	 * @param string $flag Flag determining whether the URL should be revalidated.
+	 * @return array|WP_Error Associative array containing validity result and whether the URL was revalidated, or a WP_Error on failure.
+	 */
+	public function get_url_validation( $url, $type, $flag = null ) {
+		$validity    = null;
+		$revalidated = true;
+
+		if ( self::FLAG_FORCE_REVALIDATE !== $flag ) {
+			$url_post = AMP_Validated_URL_Post_Type::get_invalid_url_post( $url );
+
+			if ( $url_post && empty( AMP_Validated_URL_Post_Type::get_post_staleness( $url_post ) ) ) {
+				$validity    = AMP_Validated_URL_Post_Type::get_invalid_url_validation_errors( $url_post );
+				$revalidated = false;
+			}
+		}
+
+		if ( self::FLAG_NO_REVALIDATE !== $flag && ( is_null( $validity ) || self::FLAG_FORCE_REVALIDATE === $flag ) ) {
+			$validity = AMP_Validation_Manager::validate_url_and_store( $url );
+		}
+
+		if ( is_wp_error( $validity ) ) {
+			return $validity;
+		} elseif ( $validity && isset( $validity['results'] ) ) {
+			$this->update_state_from_validity( $validity, $type );
+		}
```

Comment:
No need for an else, as the `return` already implies that:
```suggestion
		}

		if ( $validity && isset( $validity['results'] ) ) {
			$this->update_state_from_validity( $validity, $type );
		}
```